managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1850,9 +1854,172 @@ public void offloadFailed(ManagedLedgerException e, Object ctx) {
 
     @Override
     public void asyncOffloadPrefix(Position pos, OffloadCallback callback, Object ctx) {
-        callback.offloadFailed(new ManagedLedgerException("Not implemented"), ctx);
+        PositionImpl requestOffloadTo = (PositionImpl)pos;
+        if (!isValidPosition(requestOffloadTo)) {
+            callback.offloadFailed(new InvalidCursorPositionException("Invalid position for offload"), ctx);
+            return;
+        }
+
+        Position firstUnoffloaded;
+
+        List<LedgerInfo> ledgersToOffload = Lists.newArrayList();
+        synchronized (this) {
+            log.info("[{}] Start ledgersOffload. ledgers={} totalSize={}", name, ledgers.keySet(),
+                     TOTAL_SIZE_UPDATER.get(this));
+
+            if (STATE_UPDATER.get(this) == State.Closed) {
+                log.info("[{}] Ignoring offload request since the managed ledger was already closed", name);
+                callback.offloadFailed(new ManagedLedgerAlreadyClosedException(
+                                               "Can't offload closed managed ledger"), ctx);
+                return;
+            }
+
+            long current = ledgers.lastKey();
```

Comment:
> No, initialization always creates a new ledger and adds it to ledgers.

yes current implementation assumes creating a new ledger. but I was looking into avoiding creating a new ledger upfront for reducing metadata + better handling bookkeeper outage. so I know a lot of places in ML make this assumption, which makes my change very difficult. 

since this is a new code, we should try do null check here.